infra/batch/datasets/main.py

```diff
@@ -0,0 +1,225 @@
+        if latest_hash != file_md5_hash:
+            upload_file = True
+    else:
+        # Upload first version of dataset
+        upload_file = True
+
+    if upload_file:
+        # Upload file as latest
+        blob.upload_from_string(content)
+
+        # Upload file as upload timestamp
+        current_time = datetime.now()
+        timestamp = current_time.strftime("%Y%m%d")
+        timestamp_blob = bucket.blob(f"{stable_id}/{timestamp}.zip")
+        timestamp_blob.upload_from_string(content)
+
+        return file_md5_hash, timestamp_blob.public_url
+    return file_md5_hash, None
+
+
+def validate_dataset_version(engine, url, bucket_name, stable_id, feed_id):
+    """
+    Handles the validation of the dataset including the upload of the dataset to GCP
+    and the required database changes
+    :param engine: Database engine
+    :param url: Producer URL
+    :param bucket_name: GCP bucket name
+    :param stable_id: Feed's stable ID
+    :param feed_id: Feed's ID
+    """
+    transaction = None
+    connection = None
+    errors = ""
+    try:
+        md5_file_hash, hosted_url = upload_dataset(url, bucket_name, stable_id)
+
+        # Set up transaction for SQL updates
+        connection = engine.connect()
+        transaction = connection.begin()
+
+        # Create a new version of the dataset in the database
+        select_dataset_statement = text(f"select id, hash from gtfsdataset where latest=true and feed_id='{feed_id}'")
+        dataset_results = connection.execute(select_dataset_statement).all()
+        dataset_id = dataset_results[0][0] if len(dataset_results) > 0 else None
+        dataset_hash = dataset_results[0][1] if len(dataset_results) > 0 else None
+        print(f"Dataset ID = {dataset_id}, Dataset Hash = {dataset_hash}")
+        if dataset_id is None:
+            errors += f"[INTERNAL ERROR] Couldn't find latest dataset related to feed_id {feed_id}\n"
+            return
+
+        # Set the previous version latest field to false
+        if dataset_hash is not None and dataset_hash != md5_file_hash:
+            sql_statement = f"update gtfsdataset set latest=false where id='{dataset_id}'"
+            connection.execute(text(sql_statement))
+
+        sql_statement = f"insert into gtfsdataset (id, feed_id, latest, bounding_box, note, hash, " \
+                        f"download_date, stable_id, hosted_url) " \
+                        f"select '{str(uuid.uuid4())}', feed_id, true, bounding_box, note, " \
+                        f"'{md5_file_hash}', NOW(), stable_id, '{hosted_url}' from " \
+                        f"gtfsdataset where id='{dataset_id}'"
+
+        # In case the dataset doesn't include a hash or the dataset was deleted from the bucket,
+        # update the existing entity
+        if dataset_hash is None or dataset_hash == md5_file_hash:
+            sql_statement = f"update gtfsdataset set hash='{md5_file_hash}', hosted_url='{hosted_url}' where id='{dataset_id}'"
```

Comment:
suggestion: this is a specific use case warning us that something happened with the files or the DB integrity. I suggest logging a warning to the console to know when this use case is present.

Reply:
Agreed and updated